src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -268,6 +272,34 @@ private void CheckHeaderSignatureWithCoinstakeKernel(ProvenBlockHeader header, U
             ConsensusErrors.BadBlockSignature.Throw();
         }
 
+        /// <summary>
+        /// Checks if coinstake is spent on another chain.
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        /// <param name="context">The POS rule context.</param>
+        /// <exception cref="ConsensusException">Throws utxo not found error.</exception>
+        private void CheckIfCoinstakeIsSpentOnAnotherChain(ProvenBlockHeader header, PosRuleContext context)
+        {
+            Transaction coinstake = header.Coinstake;
+            TxIn input = coinstake.Inputs[0];
+
+            int? rewindDataIndex = this.PosParent.RewindDataIndexStore.GetAsync(input.PrevOut.Hash, (int)input.PrevOut.N).GetAwaiter().GetResult();
+            if (!rewindDataIndex.HasValue) {
+                // Not sure if we should throw this exception or another.
+                ConsensusErrors.ReadTxPrevFailed.Throw();
+            }
+
+            RewindData rewindData = this.PosParent.UtxoSet.GetRewindData(rewindDataIndex.Value).GetAwaiter().GetResult();
+            UnspentOutputs matchingUnspentUtxo =
+                rewindData.OutputsToRestore.FirstOrDefault(unspent => unspent.TransactionId == input.PrevOut.Hash);
+            if (matchingUnspentUtxo == null)
```

Comment:
this is actually problematic if we find the rewind data then we must also find the UTXO this is a bug if `matchingUnspentUtxo == null`

Reply:
agree, RewindDataIndex is the utxo we have in RewindData, so this would mean we have a store not synched (or maybe race conditions, can be possible?)
What to do then? we'd need to invalidate the cache? or leave the header stuck? or throw and stop(not ideal because it's not a peer fault but our data not synched)

Reply:
So what are we doing with this one?